File: python/test/pymidibridge/test_protocol_parallel.py
import unittest

from .mocks import *
from lib.pymidibridge.PyMidiBridge import *
    

class TestProtocolParallel(unittest.TestCase):

    def test_send_receive_parallel(self):
        # Bridge for sending data
        midi_send = MockMidiSender()
        storage_send = MockStorageProvider()        

        storage_send.read_data = {
            "foo": "OllehCasklakldmcömpöosdpo sdvsdöv sdvsd vsdvgeg etrhtrhrt ret",
            "bar": "Hello sdv wrpv wprvmevmkefv ö löevö wöorvömklfsdvnlsd dvsdvsdvsdvsssvsd"
        }

        storage_send.outputs_size = {
            "foo": len(storage_send.read_data["foo"]),
            "bar": len(storage_send.read_data["bar"])
        }

        bridge_send = PyMidiBridge(
            midi = midi_send,
            storage = storage_send,
            event_handler = MockEventHandler()
        )

        # Bridge for receiving data
        midi_receive = MockMidiSender()
        storage_receive = MockStorageProvider()        

        bridge_receive = PyMidiBridge(
            midi = midi_receive,
            storage = storage_receive,
            event_handler = MockEventHandler()
        )

        # Get a request message for the path from the receiving side
        bridge_receive.request("foo", 3)
        msg_request_foo = midi_receive.messages_sent[0]
        midi_receive.messages_sent = []

        # Get a request message for the path from the receiving side
        bridge_receive.request("bar", 4)
        msg_request_bar = midi_receive.messages_sent[0]
        midi_receive.messages_sent = []

        # Let the sending bridge receive a request message, to trigger it sending a file
        self.assertEqual(bridge_send.receive(msg_request_foo), True)        
        
        # Receive start message 
        start_msg_1 = midi_send.messages_sent.pop(0)
        data_msg_1 = midi_send.messages_sent.pop(0)
        self.assertEqual(bridge_receive.receive(start_msg_1), True)
        self.assertEqual(len(midi_receive.messages_sent), 0)

        # Let the sending bridge receive a request message, to trigger it sending a file
        self.assertEqual(bridge_send.receive(msg_request_bar), True)        
        
        # Receive start message 
        start_msg_2 = midi_send.messages_sent.pop(0)
        data_msg_2 = midi_send.messages_sent.pop(0)
        self.assertEqual(bridge_receive.receive(start_msg_2), True)
        self.assertEqual(len(midi_receive.messages_sent), 0)

        # Receive data messages
        self.assertEqual(bridge_receive.receive(data_msg_2), True)
        self.assertEqual(bridge_receive.receive(data_msg_1), True)

        self.assertEqual(len(midi_receive.messages_sent), 2)
        ack_msg_2 = midi_receive.messages_sent.pop(0)
        ack_msg_1 = midi_receive.messages_sent.pop(0)
        self.assertEqual(bridge_send.receive(ack_msg_1), True)
        self.assertEqual(bridge_send.receive(ack_msg_2), True)

        # Feed the generated MIDI messages to the receiving bridge, yielding the input data again
        while True:
            if not midi_send.messages_sent:
                # No further data
                break

            # Get next data message
            msg = midi_send.messages_sent.pop(0)
            
            # Feed data message to the receiving bridge (which must answer with ack)
            self.assertEqual(bridge_receive.receive(msg), True)

            self.assertEqual(len(midi_receive.messages_sent), 1)
            ack_msg = midi_receive.messages_sent.pop(0)

            # Feed the ack message back to the sender bridge to get the next data message
            self.assertEqual(bridge_send.receive(ack_msg), True)
            
        # Check calls
        self.assertEqual(len(storage_receive.created_handles), 2)
        self.assertEqual(storage_receive.created_handles[0].path, "foo")
        self.assertEqual(storage_receive.created_handles[0].mode, "a")
        self.assertEqual(storage_receive.created_handles[0].closed, True)

        self.assertEqual(storage_receive.created_handles[1].path, "bar")
        self.assertEqual(storage_receive.created_handles[1].mode, "a")
        self.assertEqual(storage_receive.created_handles[1].closed, True)

        # Compare results
        self.assertEqual(storage_receive.created_handles[0].write_contents, storage_send.read_data["foo"])
        self.assertEqual(storage_receive.created_handles[1].write_contents, storage_send.read_data["bar"])
    

File: js/test/htdocs/spec/test_protocol_parallel.js

describe('Protocol (Parallel Transfer)', function() {

    it('Send/receive of parallel transmissions', async function() {
		const test = new TestProtocolParallel();
        await test.testSendReceiveParallel();
    });    
});


class TestProtocolParallel {	
	
	async testSendReceiveParallel() {
        // Bridge for sending data
        const bridgeSend = new JsMidiBridge();
        
        bridgeSend.getFile = function(path_p) {
			if (path_p == "foo") {
				return "OllehCasklakldmcömpöosdpo sdvsdöv sdvsd vsdvgeg etrhtrhrt ret";
			}
			if (path_p == "bar") {
				return "Hello sdv wrpv wprvmevmkefv ö löevö wöorvömklfsdvnlsd dvsdvsdvsdvsssvsd";
			}						
		}
		
        // Bridge for receiving data
        const bridgeReceive = new JsMidiBridge();

		let finishCalls = [];
		
		bridgeReceive.onReceiveFinish = function(data) {
			finishCalls.push(data.data);
		}
		
		let messagesSentSend = [];
		bridgeSend.sendSysex = function(manufacturerId, data) {
			messagesSentSend.push({
                manufacturerId: manufacturerId,
                data: data
            });
		}

        let messagesSentReceive = [];
		bridgeReceive.sendSysex = function(manufacturerId, data) {
			messagesSentReceive.push({
                manufacturerId: manufacturerId,
                data: data
            });
		}	

        // Get a request message for the path from the receiving side
        await bridgeReceive.request("foo", 3);
        const msgRequestFoo = messagesSentReceive[0];
        messagesSentReceive = [];

        // Get a request message for the path from the receiving side
        await bridgeReceive.request("bar", 4);
        const msgRequestBar = messagesSentReceive[0];
        messagesSentReceive = [];

        // Let the bridge receive a request message, to trigger it sending a file
        expect(await bridgeSend.receive(msgRequestFoo)).toBe(true);

        // Receive start message
        const startMsg1 = messagesSentSend.shift();
        const dataMsg1 = messagesSentSend.shift();
        expect(await bridgeReceive.receive(startMsg1)).toBe(true);
        expect(messagesSentReceive.length).toBe(0);

        // Let the bridge receive a request message, to trigger it sending a file
        expect(await bridgeSend.receive(msgRequestBar)).toBe(true);

        // Receive start message
        const startMsg2 = messagesSentSend.shift();
        const dataMsg2 = messagesSentSend.shift();
        expect(await bridgeReceive.receive(startMsg2)).toBe(true);
        expect(messagesSentReceive.length).toBe(0);

        expect(await bridgeReceive.receive(startMsg2)).toBe(true);

        // Receive data messages
        expect(await bridgeReceive.receive(dataMsg2)).toBe(true);
        expect(await bridgeReceive.receive(dataMsg1)).toBe(true);

        expect(messagesSentReceive.length).toBe(2);
        const ackMsg2 = messagesSentReceive.shift();
        const ackMsg1 = messagesSentReceive.shift();
        expect(await bridgeSend.receive(ackMsg1)).toBe(true);
        expect(await bridgeSend.receive(ackMsg2)).toBe(true);

        // Feed back the generated MIDI messages to the bridge, yielding the input data again
        while(true) {
            if (messagesSentSend.length == 0) {
                // No further data
                break;
            }
            // Get next data message
            const msg = messagesSentSend.shift();

            // Feed data message to the receiving bridge (which must answer with ack)
            expect(await bridgeReceive.receive(msg)).toBe(true);

            expect(messagesSentReceive.length).toBe(1);
            const ackMsg = messagesSentReceive.shift();

            // Feed the ack message back to the sender bridge to get the next data message
            expect(await bridgeSend.receive(ackMsg)).toBe(true);
        }           

        // Be sure the results have been checked
        expect(finishCalls).toEqual(
            [                
                "Hello sdv wrpv wprvmevmkefv ö löevö wöorvömklfsdvnlsd dvsdvsdvsdvsssvsd",
                "OllehCasklakldmcömpöosdpo sdvsdöv sdvsd vsdvgeg etrhtrhrt ret"
            ]
        );
    }
}

